iina/AppDelegate.swift

```diff
@@ -296,13 +298,136 @@ class AppDelegate: NSObject, NSApplicationDelegate {
+    // command directly to mpv causing mpv and the player to shutdown before application
+    // termination is initiated.
+    var canTerminateNow = true
+    for player in PlayerCore.playerCores {
+      if !player.isShutdown {
+        canTerminateNow = false
+        break
+      }
+    }
+    if canTerminateNow {
+      Logger.log("All players have shutdown, proceeding with application termination")
+      // Tell Cocoa that it is ok to immediately proceed with termination.
+      return .terminateNow
+    }
+
+    // Shutdown of player cores involves sending the stop and quit commands to mpv. Even though
+    // these commands are sent to mpv using the synchronous API mpv executes them asynchronously.
+    // This requires IINA to wait for mpv to finish executing these commands.
+    Logger.log("Waiting for players to stop and shutdown")
+
+    // To ensure termination completes and the user is not required to force quit IINA, impose an
+    // arbitrary timeout that forces termination to complete. The expectation is that this timeout
+    // is never triggered. If a timeout warning is logged during termination then that needs to be
+    // investigated.
+    var timedOut = false
+    let timer = Timer(timeInterval: 10, repeats: false) { _ in
+      timedOut = true
+      Logger.log("Timeout during termination", level: .warning)
+      // For debugging list players that have not terminated.
+      for player in PlayerCore.playerCores {
+        let label = player.label ?? "unlabeled"
+        if !player.isStopped {
+          Logger.log("Player \(label) failed to stop", level: .warning)
+        } else if !player.isShutdown {
+          Logger.log("Player \(label) failed to shutdown", level: .warning)
+        }
+      }
+      Logger.log("Forcing application termination", level: .warning)
+      // Tell Cocoa to proceed with termination.
+      NSApp.reply(toApplicationShouldTerminate: true)
+    }
+    RunLoop.main.add(timer, forMode: .common)
+
+    // Establish an observer for a player core stopping.
+    let center = NotificationCenter.default
+    var observers: [NSObjectProtocol] = []
+    var observer = center.addObserver(forName: .iinaPlayerStopped, object: nil, queue: .main) { note in
+      guard let player = note.object as? PlayerCore else { return }
+      // Now that the player has stopped it is safe to instruct the player to terminate. IINA MUST
+      // wait for the player to stop before instructing it to terminate because sending the quit
+      // command to mpv while it is still asynchronously executing the stop command can result in a
+      // watch later file that is missing information such as the playback position. See issue #3939
+      // for details.
+      player.shutdown()
+    }
+    observers.append(observer)
+
+    // Establish an observer for a player core shutting down.
+    observer = center.addObserver(forName: .iinaPlayerShutdown, object: nil, queue: .main) { _ in
+      guard !timedOut else {
```

Comment:
I don't see the meaning of this guard and the `timeOut` variable. Soon after `timeOut` is set to true, the application is terminated forcibly, no?

Reply:
Maybe I should rephrase that "Forcing application termination" log message. The timeout tells Cocoa to proceed with termination. This does not happen immediately. Tasks on the main queue are still run after this. This guard statement insures IINA does not reply again to Cocoa telling it a second time to proceed with termination. Was worried about what Cocoa would do if IINA replied a second time. This also avoids outputing the log message that indicates a normal termination.

Eventually after IINA tells Cocoa to proceed with termination AppKit will call back to IINA, executing applicationWillTerminate. At that point IINA closes the log file. Even after that the main queue may run other tasks before the application is terminated. We know that happens because it was causing crashes when a task tried to log a message after the log file was closed.

Maybe the log message should be: "Timeout. Proceeding with application termination"?

Reply:
Okay, got that. I also think is reasonable to change the log message.